src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,254 @@
+    // parameterization
+    for (unsigned int i = 0; i < num_segments; i++)
+    {
+        // Get the 't' value corresponding to the current arc length (to be used for
+        // further computing)
+        double t = shared_polynomial2d_getTValueAtArcLengthOrder3(
+            path, i * arc_segment_length, arc_length_param);
+
+        // Get the X and Y position at the 't' value defined by the arc length
+        traj_elements[i].position = shared_polynomial2d_getValueOrder3(path, t);
+
+
+        // Create the polynomial representing path curvature
+        //                                              1
+        //                              ---------------------------------
+        //                                     abs(x'y'' - y'x'')
+        //        radius of curvature =      ----------------------
+        //                                     (x'^2 + y'^2)^(3/2)
+        //
+        const double numerator =
+            fabs(shared_polynomial1d_getValueOrder2(first_deriv.x, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.y, t) -
+                 shared_polynomial1d_getValueOrder2(first_deriv.y, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.x, t));
+        const double denominator =
+            pow(pow(shared_polynomial1d_getValueOrder2(first_deriv.x, t), 2) +
+                    pow(shared_polynomial1d_getValueOrder2(first_deriv.y, t), 2),
+                3.0 / 2.0);
+
+        const double radius_of_curvature = 1 / (numerator / denominator);
+
+        max_allowable_speed_profile[i] =
+            sqrt(max_allowable_acceleration * radius_of_curvature);
+    }
+
+    // First point on the velocity profile is always the current speed
+    velocity_profile[0] = initial_speed;
+
+    // Loop through the path forwards to ensure that the maximum velocity limit defined by
+    // the curvature is not breached by constant max acceleration of the robot (if it is,
+    // pull down the acceleration)
+    for (unsigned int i = 1; i < num_segments; i++)
+    {
+        // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
+        const double temp_vel = sqrt(pow(velocity_profile[i - 1], 2) +
+                                     2 * arc_segment_length * max_allowable_acceleration);
+
+        // If the new speed is greater than the max allowable, reduce it to the max
+        if (temp_vel > max_allowable_speed_profile[i])
+        {
+            velocity_profile[i] = max_allowable_speed_profile[i];
+        }
+
+        else
+        {
+            // If the new speed is greater than the maximum limit of the robot, reduce it
+            // to the limit
+            if (temp_vel >= max_allowable_speed)
+            {
+                velocity_profile[i] = max_allowable_speed;
+            }
+            else
+            {
+                velocity_profile[i] = temp_vel;
+            }
+        }
```

Comment:
I think we can simplify the `if` here a bit:
```
        // If the new speed is greater than the max allowable, reduce it to the max
        if (temp_vel > max_allowable_speed_profile[i])
        {
            velocity_profile[i] = max_allowable_speed_profile[i];
        }
        // If the new speed is greater than the maximum limit of the robot, reduce it
        // to the limit
        else if (temp_vel >= max_allowable_speed)
        {
            velocity_profile[i] = max_allowable_speed;
        }
        else
        {
            velocity_profile[i] = temp_vel;
        }
```

Reply:
Very true, simplified